flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -1101,11 +1107,70 @@ class ExprLowering {
         si.box);
   }
 
+  fir::ExtendedValue genArrayCoorOp(const Fortran::lower::SymbolBox &si,
+                                    const Fortran::evaluate::ArrayRef &aref) {
+    auto loc = getLoc();
+    auto addr = si.getAddr();
+    auto arrTy = fir::dyn_cast_ptrEleTy(addr.getType());
+    auto eleTy = arrTy.cast<fir::SequenceType>().getEleTy();
+    auto refTy = builder.getRefType(eleTy);
+    auto idxTy = builder.getIndexType();
+    auto genWithShape = [&](const auto &arr) -> mlir::Value {
+      mlir::Type shapeType =
+          fir::ShapeType::get(builder.getContext(), arr.getExtents().size());
+      auto shapeOp =
+          builder.create<fir::ShapeOp>(loc, shapeType, arr.getExtents());
```

Comment:
It looks like the SahpeOp is independent of the subscripts, Should the ShapeOp be computed/propagated when creating the extended value instead of at every reference ? (that may be a bad idea).

Reply:
I think its also fine to start the easiest/correct way and recreate the shape every time as you are and to later see if this can be safely improved and it's worth it.

Reply:
Hopefully, redundant ShapeOps will get merged by the CSE pass.